internal/datastore/common/changes.go

```diff
@@ -4,99 +4,221 @@ import (
 	"context"
 	"sort"
 
+	"golang.org/x/exp/maps"
+
 	log "github.com/authzed/spicedb/internal/logging"
 	"github.com/authzed/spicedb/pkg/datastore"
 	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+	"github.com/authzed/spicedb/pkg/spiceerrors"
 	"github.com/authzed/spicedb/pkg/tuple"
 )
 
-// Changes represents a set of tuple mutations that are kept self-consistent
+const (
+	nsPrefix     = "n$"
+	caveatPrefix = "c$"
+)
+
+// Changes represents a set of datastore mutations that are kept self-consistent
 // across one or more transaction revisions.
 type Changes[R datastore.Revision, K comparable] struct {
 	records map[K]changeRecord[R]
 	keyFunc func(R) K
+	content datastore.WatchContent
 }
 
 type changeRecord[R datastore.Revision] struct {
-	rev          R
-	tupleTouches map[string]*core.RelationTuple
-	tupleDeletes map[string]*core.RelationTuple
+	rev                R
+	tupleTouches       map[string]*core.RelationTuple
+	tupleDeletes       map[string]*core.RelationTuple
+	definitionsChanged map[string]datastore.SchemaDefinition
+	namespacesDeleted  map[string]struct{}
+	caveatsDeleted     map[string]struct{}
 }
 
 // NewChanges creates a new Changes object for change tracking and de-duplication.
-func NewChanges[R datastore.Revision, K comparable](keyFunc func(R) K) Changes[R, K] {
-	return Changes[R, K]{
-		make(map[K]changeRecord[R], 0),
-		keyFunc,
+func NewChanges[R datastore.Revision, K comparable](keyFunc func(R) K, content datastore.WatchContent) *Changes[R, K] {
+	return &Changes[R, K]{
+		records: make(map[K]changeRecord[R], 0),
+		keyFunc: keyFunc,
+		content: content,
 	}
 }
 
-// AddChange adds a specific change to the complete list of tracked changes
-func (ch Changes[R, K]) AddChange(
+// IsEmpty returns if the change set is empty.
+func (ch *Changes[R, K]) IsEmpty() bool {
+	return len(ch.records) == 0
+}
+
+// AddRelationshipChange adds a specific change to the complete list of tracked changes
+func (ch *Changes[R, K]) AddRelationshipChange(
 	ctx context.Context,
 	rev R,
 	tpl *core.RelationTuple,
 	op core.RelationTupleUpdate_Operation,
-) {
+) error {
+	if ch.content&datastore.WatchRelationships != datastore.WatchRelationships {
```

Comment:
we don't have test coverage for this - make sure a relationship change is not reported if WatchRelationships is not requested